MakeHist/mumu.C

```diff
@@ -119,58 +127,53 @@ void mumu::Loop(TString sample_name)
+                                   jeteta_hist->Fill(jet_eta[NJ]);
 
-                                   //di-lepton information (delta-Phi, delta-R, dilepton mass)
-                                   Dil_dPhi[0] = dil_dPhi;
-                                   Dil_dR[0] = dil_dR;
-                                   Dil_mass[0] = dil_mass;
-
-                                   //Missing ET
-                                   MET[0] = met_pt;
-
-                                   //Save all jet information 
-
-                                   TLorentzVector Jet;
-                                   Jet.SetPtEtaPhiE(jet_pt[NJ],jet_eta[NJ],jet_phi[NJ],jet_E[NJ]);
-                                   Jets.push_back(Jet);
-
-                                   //Number of jet
-                                   Njet ++;
 
                                    if(jet_btag[NJ] >= 0.8484) { //b-tagging(WP=0.8484)
+                                       //b-jet pT and eta
+                                       bjetpT_hist->Fill(jet_pt[NJ]);
+                                       bjeteta_hist->Fill(jet_eta[NJ]);
                                        //Save b-jet information
                                        TLorentzVector bJet;
                                        bJet.SetPtEtaPhiE(jet_pt[NJ],jet_eta[NJ],jet_phi[NJ],jet_E[NJ]);
                                        bJets.push_back(bJet);
+                                   }
+
+                                   else if(jet_pt[NJ] > 30){//non_bjet pT > 30
+                                       //non b-jet pT and eta
+                                       NonbjetpT_hist->Fill(jet_pt[NJ]);
+                                       Nonbjeteta_hist->Fill(jet_eta[NJ]);
+                                       //Save non b-jet information
+                                       TLorentzVector non_bJet;
+                                       non_bJet.SetPtEtaPhiE(jet_pt[NJ],jet_eta[NJ],jet_phi[NJ],jet_E[NJ]);
+                                       non_bJets.push_back(non_bJet);
 
-                                       //Number of b-jet
-                                       Nbjet ++;
                                    }
                                }
                            }
                        }// b jet pT > 20 GeV , non-b jet pT > 30 GeV
 
-                       njet_hist->Fill(Njet);
-                       if(Njet < 2) continue;//Number of jet >=2
-                       
-                       nbjet_hist->Fill(Nbjet);
-                       if(Nbjet < 1) continue;// >= 1 b-jet
+                       nNonbjet_hist->Fill(non_bJets.size());
+                       njvsnbj_hist->Fill(non_bJets.size(),bJets.size());
+                       nbjet_hist->Fill(bJets.size());
+
+                       //Case 1 = Case 3 + Case 4 + Case 5
+                       if(bJets.size() < 1 || bJets.size()+non_bJets.size() < 2) continue;//N_jet >= 2, N_bjets >= 1 (Case 1)
```

Comment:
Okay, that's your preselection, because what's skipped here is in neither the exactly 1b selection, nor in the at least 1 b N_b+N_notb=2 selection.
Now to distinguish the two, you could add after that line e.g.
`bool MonoBselection=false;
bool TwoJetSelection=false;
if(bJets.size()+non_bJets.size()==1) MonoBSelection=true;
else if(bJets.size()+non_bJets.size()==2) TwoJetSelection=true;`
and then do your plotting of variables for histograms for each selection according to which selection bool is true. E.g. for mass:
`if(MonoBSelection) MonoBSelection_dimuonMass->Fill(mass,weight);
if(TwoJetSelection) TwoJetSelection_dimuonMass->Fill(mass,weight);`
Right now, it seems there is still some manual switching that has to happen for you to produce either selection. Producing both at once is less error-prone.

Reply:
Thank you for your feedback.

Does it mean to draw MonoBselection plots and TwoJetSelection plots(e.g. leppT, jetpT, SBM..., etc.) for all plots after # of (b)jets cut?
Additionally, TwoJetSelection may also be true when N_bjets = 0. (When N_bjets = 0, N_nonbjets = 2)
Is this what you intend?
Or did you intend N_jets = 2, N_bjets = 1 or 2 ?

Reply:
@hyeahyun No, I was assuming that the line `if(bJets.size() < 1 || bJets.size()+non_bJets.size() < 2) continue;` remains unaltered before defining and later using either bool that determines which of the two selections applies to the current event.
That way, there can never be 0 bjets.

Reply:
Okay, I understood your means.
Could you answer the first question?
Now, the number of plots after # of (b)jets cut is about 20.
If I do this all for MonoBselection plots and TwoJetSelection plots, I draw about 60 plots.

I think this has a little inefficiency...

Reply:
@hyeahyun You could formulate all histograms you declare as elements in `std::vector<TH1F*>` objects. You'd have to change the definition of a TH1F though such that the physical name differs. e.g. 
`std::vector<TH1F*> histogram1;
for(unsigned selection=0; selection>1; selection++){
  histogram1.push_back((TH1F*) new TH1F(TString::Format("histname_%d",selection), "histtitle;x-axis descriptor;y-axis descriptor", 1,0.,1.));
}`

And at the end, you'd get vectors you could run over or write about to dump them into a file. You can also change the original bool selection to decide by an int, e.g. -1 -> continue, 0 -> MonoBselection, 1-> TwoBselection to directly address the histogram that has to be filled.
`if(bJets.size()<1) continue;
int selectInt=-1;
if(bJets.size()+non_bJets.size()==1) selectInt=0;
else if(bJets.size()+non_bJets.size()==2) selectInt=1;
if(selectInt<0) continue;
histogram[selectInt]->Fill(variable,weight);`
That way you can keep the code concise and still produce all plots. Does that help you with the inefficiency concern?